python/smqtk/algorithms/saliency/sbsm.py

```diff
@@ -0,0 +1,266 @@
+from PIL import Image
+import os
+from sklearn.metrics.pairwise import euclidean_distances
+import six
+from smqtk.algorithms import SmqtkAlgorithm
+from smqtk.algorithms.saliency import SaliencyBlackbox,ImageSaliencyAugmenter
+from smqtk.representation.data_element.memory_element import DataMemoryElement
+
+try:
+    import numpy as np
+    from tqdm import tqdm
+    import copy
+except ImportError as ex:
+    logging.getLogger(__name__).warning("Failed to import numpy/tqdm \
+         /copy module: %s", str(ex))
+    np = None
+    tqdm = None
+    copy = None
+
+
+class SBSM_SaliencyBlackbox (SaliencyBlackbox):
+    """
+    SBSM_SaliencyBlackbox function that yields some floating point 
+    scalar value for a given masked base image descriptor element that 
+    signifies the proximity between the query image and masked image
+    descriptors, used by class implementations of 'ImageSaliencyMapGenerator'.
+    """
+
+    def __init__(self, query_f, base_descr):
+        """
+        :param query_f: Feature of query image
+        :param type: smqtk.representation.DescriptorElement
+        :param base_descr: Base image descriptor
+        :param type: smqtk.representation.DescriptorElement
+        """    
+
+        self.query_f = query_f        
+        self.base_descr = base_descr
+
+    @classmethod
+    def is_usable(cls):
+        """
+        Check whether this implementation is available for use.
+        Required valid presence of query image feature 
+        and base image descriptor
+        :return:
+            Boolean determination of whether this implementation is usable.
+        :rtype: bool
+        """
+
+        valid = True
+        if not valid:
+            cls.get_logger().debug("six python module cannot be imported")
+        return valid
+                
+    @classmethod
+    def from_iqr_session(cls, iqrs, descr_gen, base_image): 
+        """
+        Create an ``SaliencyBlackbox`` instance from iqrs session,
+        descriptor generator and base_image.
+        :param iqrs:`smqtk.iqr.IqrSession` instance.
+        :param type: smqtk.iqr.IqrSession
+        :param descr_gen: The descriptor generator used by smqtk.
+        :param type: smqtk.algorithms.DescriptorGenerator.
+        :param base_image: The Base image for which we need to
+        calculate a saliency map.
+        :param type: PIL Image of the base image.  
+        :return: A new instance of a class implementing the
+            ``SaliencyBlackbox`` class.
+        :rtype: SaliencyBlackbox
+        """
+
+        assert iqrs
+        if (len(iqrs.external_positive_descriptors)) != 1:
+            raise NotImplementedError("Saliency generation in class``{}`` "
```

Comment:
This should probably not be a ``NotImplementedError`` as that is meant to indicate that a function/method is not implemented at all. A ``ValueError`` may be more fitting here as the issue is with the ``IqrSession`` instance provided.

Reply:
Changed to ValueError.